File: binary search trees/delete a node.cpp

class Solution {
public:
    // get min value in the right subtree
    int getMinValue(TreeNode * root){
        if (root == nullptr){
            return -1;
        }
        if (root->left == nullptr){
            return root->val;
        }
        // min value only exists in left subtree
        return getMinValue(root->left);
    }
    TreeNode* deleteNode(TreeNode* root, int key) {
        if (root == nullptr){
            return root;
        }
        if (key < root->val) root->left = deleteNode(root->left, key);
        else if(key > root->val) root->right = deleteNode(root->right, key);
        else {
            //we got the matched node
            // case 1: no child
            if (root->left == nullptr && root->right == nullptr){
                delete root;
                root = nullptr;
            }
            // case 2: one child
            // right subtree is null
            else if (root->right == nullptr){
                TreeNode * temp = root->left;
                delete root;
                root = temp;
            }
            // left subtree is null
            else if (root->left == nullptr){
                TreeNode * temp = root->right;
                delete root;
                root = temp;
            }
            // case 3: two child
            else {
                // find min value in right subtree
                int temp = getMinValue(root->right);
                // set the root node value to temp
                root->val = temp;
                // aftering setting root value to temp
                // we have duplicate value in the subtree, we just have to delete that 
                // node in the right subtree
                root->right = deleteNode(root->right, temp);
            }
        }
        return root;
    }
};
// Another way to solve case 3
 // take the max value from the right subtree and solve in the same way

File: binary search trees/delete a node.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def minValueFromRightSubtree(self, root):
        if root is None:
            return -1
        if root.left is None:
            return root.val
        return self.minValueFromRightSubtree(root.left)

    def deleteNode(self, root: Optional[TreeNode], key: int) -> Optional[TreeNode]:
        if root is None:
            return root
        if key < root.val:
            root.left = self.deleteNode(root.left, key)
        elif key > root.val:
            root.right = self.deleteNode(root.right, key)
        else:
            # case 1: no child 
            if root.left == None and root.right == None:
                del root
                root = None
            # case 2: one child
            elif root.left is None:
                temp = root.right
                del root
                root = temp
            elif root.right is None:
                temp = root.left
                del root
                root = temp
            else:
                minValue = self.minValueFromRightSubtree(root.right)
                root.val = minValue
                root.right = self.deleteNode(root.right, minValue)
        return root
